crates/spin-wasip3-http/src/lib.rs

```diff
@@ -0,0 +1,526 @@
+        req.set_scheme(scheme.as_ref())
+            .map_err(|()| types::ErrorCode::HttpProtocolError)?;
+
+        req.set_authority(parts.uri.authority().map(|a| a.as_str()))
+            .map_err(|()| types::ErrorCode::HttpRequestUriInvalid)?;
+
+        req.set_path_with_query(parts.uri.path_and_query().map(|pq| pq.as_str()))
+            .map_err(|()| types::ErrorCode::HttpRequestUriInvalid)?;
+
+        wit_bindgen::spawn(async move {
+            let mut body = std::pin::pin!(body);
+            _ = body_writer.forward_http_body(&mut body).await;
+        });
+
+        Ok(req)
+    }
+}
+
+impl FromRequest for types::Request {
+    fn from_request(req: types::Request) -> HttpResult<Self> {
+        Ok(req)
+    }
+}
+
+impl<T: FromRequest> FromRequest for http::Request<T> {
+    fn from_request(req: types::Request) -> HttpResult<Self> {
+        let uri = {
+            let mut builder = http::Uri::builder();
+            if let Some(scheme) = req.get_scheme() {
+                builder = builder.scheme(scheme_from_wasi(scheme)?);
+            }
+            if let Some(authority) = req.get_authority() {
+                builder = builder.authority(authority);
+            }
+            if let Some(path_and_query) = req.get_path_with_query() {
+                builder = builder.path_and_query(path_and_query);
+            }
+            builder
+                .build()
+                .map_err(|_| types::ErrorCode::HttpRequestUriInvalid)?
+        };
+
+        let mut builder = http::Request::builder()
+            .method(method_from_wasi(req.get_method())?)
+            .uri(uri);
+
+        if let Some(options) = req.get_options().map(RequestOptionsExtension) {
+            builder = builder.extension(options);
+        }
+
+        for (k, v) in req.get_headers().copy_all() {
+            builder = builder.header(k, v);
+        }
+
+        let body = T::from_request(req)?;
+
+        builder.body(body).map_err(to_internal_error_code) // TODO: downcast to more specific http error codes
+    }
+}
+
+impl<T: FromResponse> FromResponse for http::Response<T> {
+    fn from_response(resp: types::Response) -> HttpResult<Self> {
+        let mut builder = http::Response::builder().status(resp.get_status_code());
+
+        for (k, v) in resp.get_headers().copy_all() {
+            builder = builder.header(k, v);
+        }
+
+        let body = T::from_response(resp)?;
+        builder.body(body).map_err(to_internal_error_code) // TODO: downcast to more specific http error codes
+    }
+}
+
+impl FromRequest for () {
+    fn from_request(_req: types::Request) -> HttpResult<Self> {
+        Ok(())
+    }
+}
+
+impl IntoResponse for types::Response {
+    fn into_response(self) -> HttpResult<types::Response> {
+        Ok(self)
+    }
+}
+
+impl<T: http_body::Body> IntoResponse for (http::StatusCode, T) {
+    fn into_response(self) -> HttpResult<types::Response> {
+        unreachable!()
```

Comment:
Is this intentional?  Using `unreachable!` rather than `todo!()` makes me assume so but maybe add a comment on _why_ it is unreachable?

Reply:
Oh this one slipped by! Ill fix, thanks for calling it out!